src/schema.rs

```diff
@@ -33,18 +33,35 @@ pub trait SchemaType {
     fn get_type() -> crate::schema::Type;
 }
 
-/// Contains all the contract schemas for a V0 module
+/// Contains all the contract schemas for a smart contract module V0.
+///
+/// When embedded into a smart contract module, name the custom section
+/// `concordium-schema-v1`.
 #[derive(Debug, Clone)]
 pub struct ModuleV0 {
     pub contracts: BTreeMap<String, ContractV0>,
 }
 
-/// Contains all the contract schemas for a V1 module
+/// Contains all the contract schemas for a smart contract module V1.
+///
+/// When embedded into a smart contract module, name the custom section
+/// `concordium-schema-v2`.
 #[derive(Debug, Clone)]
 pub struct ModuleV1 {
     pub contracts: BTreeMap<String, ContractV1>,
 }
 
+/// Represents every schema module starting from smart contract module V1.
+///
+/// The serialization of this type includes the versioning information.
+///
+/// When embedded into a smart contract module, name the custom section
+/// `concordium-schema`.
+#[derive(Debug, Clone)]
+pub enum VersionedModuleSchema {
+    V1(ModuleV1),
```

Comment:
So in principle we could also support this for V0 modules. It would have pleasing symmetry.

Reply:
Then the situation isn't that confusing. V0 modules can have custom sections concordium-schema-v1 and concordium-schema, V1 modules can have concordium-schema-v2 and concordium-schema.

The latter one contains the versioning inside, and there are some restrictions on which schema can apply to which module version.

Reply:
This is now implemented